src/Microsoft.AspNetCore.SignalR.Protocols.MsgPack/Internal/Protocol/MessagePackHubProtocol.cs

```diff
@@ -544,14 +570,118 @@ private static object DeserializeObject(Unpacker unpacker, Type type, string fie
+            }
+
+            private static class Cache<T>
+            {
+                public static readonly IMessagePackFormatter<T> Formatter;
+
+                static Cache()
+                {
+                    foreach (var resolver in Resolvers)
+                    {
+                        Formatter = resolver.GetFormatter<T>();
+                        if (Formatter != null)
+                        {
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+
+        internal class CombinedResolvers : IFormatterResolver
+        {
+            private readonly IList<IFormatterResolver> _resolvers;
+
+            public CombinedResolvers(IList<IFormatterResolver> resolvers)
+            {
+                _resolvers = resolvers;
+            }
+
+            public IMessagePackFormatter<T> GetFormatter<T>()
+            {
+                foreach (var resolver in _resolvers)
+                {
+                    var formatter = resolver.GetFormatter<T>();
+                    if (formatter != null)
+                    {
+                        return formatter;
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        internal class CustomDateTimeResolver : IFormatterResolver
+        {
+            public static readonly IFormatterResolver Instance = new CustomDateTimeResolver();
+
+            private static class FormatterCache<T>
+            {
+                public static readonly IMessagePackFormatter<T> formatter;
+
+                static FormatterCache()
+                {
+                    if (typeof(T) == typeof(DateTime))
+                    {
+                        formatter = (IMessagePackFormatter<T>)CustomDateTimeFormatter.Instance;
+                        return;
+                    }
+                    formatter = null;
+                }
+            }
+
+            private CustomDateTimeResolver()
+            {
+            }
+
+            public IMessagePackFormatter<T> GetFormatter<T>()
+            {
+                return FormatterCache<T>.formatter;
+            }
+        }
+
+        private class CustomDateTimeFormatter : IMessagePackFormatter<DateTime>
+        {
+            public static readonly CustomDateTimeFormatter Instance = new CustomDateTimeFormatter();
+
+            CustomDateTimeFormatter()
+            {
+            }
+
+            public int Serialize(ref byte[] bytes, int offset, DateTime value, IFormatterResolver formatterResolver)
+            {
+                value = DateTime.SpecifyKind(value, DateTimeKind.Utc);
+                return MessagePackBinary.WriteDateTime(ref bytes, offset, value);
+            }
+
+            public DateTime Deserialize(byte[] bytes, int offset, IFormatterResolver formatterResolver, out int readSize)
+            {
+                return MessagePackBinary.ReadDateTime(bytes, offset, out readSize);
```

Comment:
What's the `Kind` value for this? `Unspecified`? That's what I'm expecting.

Reply:
I think so, didn't check in the debugger

Reply:
It reads as UTC

Reply:
We should use SpecifyKind to force the kind to Unspecified then. We don't know the kind, so we should model that properly.;